tests/test_losses.py

```diff
@@ -3,114 +3,185 @@
 """Test the PyKEEN custom loss functions."""
 
 import unittest
-from typing import Any, Mapping, Optional, Type
+from typing import Mapping
 
 import torch
-from torch.nn import functional
 
-from pykeen.losses import BCEAfterSigmoidLoss, CrossEntropyLoss, Loss, NSSALoss, SoftplusLoss
-from pykeen.pipeline import PipelineResult, pipeline
-
-
-class _LossTests:
+from pykeen.losses import (
+    BCEAfterSigmoidLoss,
+    BCELoss,
+    CrossEntropyLoss,
+    Loss,
+    MSELoss,
+    MarginRankingLoss,
+    NSSALoss,
+    PairwiseLoss,
+    PointwiseLoss,
+    SetwiseLoss,
+    SoftplusLoss,
+    losses,
+)
+from pykeen.pipeline import pipeline
+from tests.base import GenericTest, TestsTest
+
+
+class _LossTests(GenericTest[Loss]):
     """Base unittest for loss functions."""
 
-    #: The class
-    cls: Type[Loss]
-
-    #: Constructor keyword arguments
-    kwargs: Optional[Mapping[str, Any]] = None
-
-    #: The loss instance
-    instance: Loss
-
     #: The batch size
     batch_size: int = 3
 
-    def setUp(self) -> None:
-        """Initialize the instance."""
-        kwargs = self.kwargs
-        if kwargs is None:
-            kwargs = {}
-        self.instance = self.cls(**kwargs)
+    #: Which training loops are supported
```

Comment:
maybe we could have two class variables in the `Loss` class called `supports_lcwa` and `supports_slcwa` that are set to true, and any subclass can feel free to explicitly set them to false

Reply:
I thought about the same. I also saw that we have a `loss_blacklist` in the `TrainingLoop`

https://github.com/pykeen/pykeen/blob/93816a70cf21b1773feed76c83fb3d98a229db02/src/pykeen/training/training_loop.py#L69

which stores the same information.

Reply:
hmmmm I think it would be better to store this information in the loss so it would be possible to independently add new loss functions of the training loops